File: code/src/main/scala/thoughtworks/cleancode/deduplication/scala/People.scala
package thoughtworks.cleancode.deduplication.scala

case class Person(age:Int, name:String, gender:String)

class People {
  val persons: List[Person] = List(Person(22, "zhangyi", "male"));

}

object People {
  val persons: List[Person] = List(Person(22, "zhangyi", "male"));

  def find(predicate:Person => Boolean):List[Person] = {
    var result:List[Person] = Nil
    for (p <- persons) {
      if (predicate(p)) result = p::result
    }
    result
  }


  val personWith23age = find(_.age == 23)
  val personsBelow23age = find(_.age < 23)
  val personsWithName = find(_.name.equals("zhangyi"))

  val resultByAge = persons.map(p => (p.name, p.age))
    .filter(_._1 == "zhangyi")
  persons.takeWhile(p => p.name == "zhangyi")
}

File: src/main/java/zhangyi/cleancode/deduplication/People.java
package zhangyi.cleancode.deduplication;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public class People {
    private List<Person> persons = new ArrayList<Person>();

    public List<Person> findByName(String name) {
        return findBy(equalsName(name));
    }

    public List<Person> findByAge(int age) {
        return findBy(equalsAge(age));
    }

    public List<Person> findByGender(Gender gender) {
        return findBy(equalsGender(gender));
    }

    public static Predicate<Person> equalsName(String name) {
        return p -> p.getName().equals(name);
    }

    public static Predicate<Person> equalsAge(int age) {
        return p -> p.getAge() == age;
    }

    public static Predicate<Person> equalsGender(Gender gender) {
        return p -> p.getGender().equals(gender);
    }

    public List<Person> findBy(Predicate<Person> predicate) {
        return persons.stream().filter(predicate).collect(Collectors.toList());
    }

    public void add(Person person) {
        persons.add(person);
    }
}
